src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1021,12 +1024,20 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     var lazySemanticModel = body == null ? null : new Lazy<SemanticModel>(() =>
                     {
                         var syntax = body.Syntax;
-                        var semanticModel = (CSharpSemanticModel)_compilation.GetSemanticModel(syntax.SyntaxTree);
-                        var memberModel = semanticModel.GetMemberModel(syntax);
-                        if (memberModel != null)
+                        var semanticModel = (SyntaxTreeSemanticModel)_compilation.GetSemanticModel(syntax.SyntaxTree);
+
+                        if (forSemanticModel.Syntax != null)
                         {
-                            memberModel.UnguardedAddBoundTreeForStandaloneSyntax(syntax, body);
+                            semanticModel.GetOrAddModel((CSharpSyntaxNode)forSemanticModel.Syntax,
+                                                        (rootSyntax) =>
+                                                        {
```

Comment:
To avoid the closure *always* begin created, even when this enclosing `if` statement is not taken, it might be good to copy the members of `forSemanticModel` into one or more fresh variables which are used inside the lambda. Then we'd only get a closure created when the `if` is taken, which should not occur during normal compilation.

Reply:
Or at least, no closure when not running analyzers.

---
In reply to: [170042730](https://github.com/dotnet/roslyn/pull/24918#discussion_r170042730) [](ancestors = 170042730)